internal/sources/api_upstream.go

```diff
@@ -73,20 +98,143 @@ func (h *upstreamAPIHandler) Validate(ctx context.Context, endpoint string) erro
+
+	// Convert to UpstreamRegistry format
+	upstreamReg := h.buildUpstreamRegistry(servers)
+
+	// Calculate hash
+	hash, err := h.calculateHash(upstreamReg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate hash: %w", err)
+	}
+
+	// Return as FetchResult
+	return NewFetchResult(upstreamReg, hash, config.SourceFormatUpstream), nil
 }
 
 // CurrentHash returns the current hash of the API response
-// Phase 2: Not yet implemented
-func (*upstreamAPIHandler) CurrentHash(_ context.Context, _ *config.RegistryConfig) (string, error) {
-	return "", fmt.Errorf("upstream MCP Registry API support not yet implemented (Phase 2)")
+// TODO: Optimize this - could use HEAD request, ETag header, or last-modified header
+// For now, perform full fetch to get hash (simple but consistent with git/file handlers)
+func (h *upstreamAPIHandler) CurrentHash(ctx context.Context, regCfg *config.RegistryConfig) (string, error) {
+	result, err := h.FetchRegistry(ctx, regCfg)
+	if err != nil {
+		return "", err
+	}
+	return result.Hash, nil
 }
 
-// TODO Phase 2 implementation:
-// - Implement pagination support with cursor handling
-// - Fetch /v0/servers with limit/cursor parameters
-// - Convert upstream ServerDetail format to ToolHive ImageMetadata
-// - Handle version-specific endpoints /v0/servers/{name}/versions
-// - Support authentication (Bearer tokens, API keys)
+// fetchAllServers performs paginated fetching and returns all ServerJSON objects
+func (h *upstreamAPIHandler) fetchAllServers(ctx context.Context, baseURL string) ([]v0.ServerJSON, error) {
+	logger := log.FromContext(ctx)
+	allServers := []v0.ServerJSON{}
+	cursor := ""
+	pageCount := 0
+
+	for {
+		pageCount++
+
+		// Security: Prevent infinite pagination loops
+		if pageCount > maxPaginationPages {
+			return nil, fmt.Errorf(
+				"pagination exceeded maximum pages (%d), possible infinite loop or malicious upstream",
+				maxPaginationPages,
+			)
+		}
+
+		// Build URL with pagination
+		requestURL := fmt.Sprintf("%s/v0.1/servers?limit=100", baseURL)
+		if cursor != "" {
+			// Security: URL-encode cursor to prevent injection attacks
+			requestURL = fmt.Sprintf("%s&cursor=%s", requestURL, url.QueryEscape(cursor))
+		}
+
+		logger.V(1).Info("Fetching page", "page", pageCount, "url", requestURL)
+
+		// Fetch page
+		data, err := h.httpClient.Get(ctx, requestURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch page %d: %w", pageCount, err)
+		}
+
+		// Parse response
+		var response v0.ServerListResponse
+		if err := json.Unmarshal(data, &response); err != nil {
+			return nil, fmt.Errorf("failed to parse response page %d: %w", pageCount, err)
+		}
+
+		logger.V(1).Info("Parsed page", "page", pageCount, "serversInPage", len(response.Servers))
+
+		// Security: Prevent memory exhaustion from too many servers
+		if len(allServers)+len(response.Servers) > maxServers {
+			return nil, fmt.Errorf("total servers (%d) would exceed maximum (%d), possible DoS attack",
```

Comment:
Not sure if its a DoS attack given the registry server is making the call, but it may definitely result in an "out of service" issue.